src/targets/maven.ts

```diff
@@ -187,31 +193,140 @@ export class MavenTarget extends BaseTarget {
+      // party plugin for releases.
+      await retrySpawnProcess(this.mavenConfig.gradleCliPath, [
+        'closeAndReleaseRepository',
+      ]);
+    });
+  }
 
-    // Maven central is very flaky, so retrying with an exponential delay in
-    // in case it fails.
-    // TODO: close the repository by doing the requests from Craft
-    // What the gradle plugin does is a / some requests to close the repo, see
-    // https://github.com/vanniktech/gradle-maven-publish-plugin/tree/master/src/main/kotlin/com/vanniktech/maven/publish/nexus
-    // If Craft did those requests, it wouldn't be necessary to rely on a third
-    // party plugin for releases.
-    await retrySpawnProcess(this.mavenConfig.gradleCliPath, [
-      'closeAndReleaseRepository',
-    ]);
-    await this.deleteUserGradlePropsFile();
+  /**
+   * Encapsulates the given callback in a context manager. Maven CLI requires
+   * Maven Central credentials to be available in the gradle properties, which
+   * should be deleted to not expose them. The context manager makes a snapshot
+   * of the gradle properties, adds the credentials, runs the callback, and
+   * restores the snapshot.
+   *
+   * @param cb Callback to encapsulate.
+   */
+  private async withGradleProps(
+    cb: CallbackFunction
+  ): ReturnType<CallbackFunction> {
+    await withTempDir(async dir => {
+      const snapshotPath = await this.makeGradlePropsSnapshot(dir);
+      // Overwrite the file, instead of appending the content, because:
+      // 1. The target must get all the config by itself, without relying on
+      // local config files.
+      // 2. Appending more config to the gradle config file may imply a
+      // misconfiguration (such as duplicated or conflicting keys).
+      await this.createUserGradlePropsFile();
+
+      try {
+        return await cb();
+      } finally {
+        if (snapshotPath) {
+          await this.restoreGradleProps(snapshotPath);
+        } else {
+          this.logger.debug('Deleting temporary gradle properties file...');
+          this.deleteUserGradlePropsFile();
+        }
+      }
+    });
   }
 
   /**
-   * Creates the required user's `gradle.properties` file.
+   * Makes a snapshot of the gradle properties file, if exists.
    *
-   * If there's an existing one, it's overwritten.
-   * TODO: control when it's overwritten with an option.
+   * @param snapshotDir Directory to place the snapshot in.
+   * @returns The path to the snapshot in the given directory,
+   *  or `undefined` if the snapshot wasn't created.
+   */
+  public async makeGradlePropsSnapshot(
+    snapshotDir: string
+  ): Promise<string | undefined> {
+    const propsExpectedPath = this.getGradlePropsPath();
+    try {
+      accessSync(propsExpectedPath, fsConstants.R_OK);
+    } catch (error) {
+      this.logger.debug(
+        'Did not find a gradle properties file in: ',
+        propsExpectedPath
+      );
+      return undefined;
```

Comment:
```suggestion
      return null;
```

`null` might be a better option here to signify intentional empty return value.